docs/source/deprecation.md

```diff
@@ -1,17 +1,458 @@
-## Deprecation
+# Deprecation
+
+The deprecation process is here to make (your) life of developers easier.
+It's not possible to write perfect solution for everything and as the project
+is evolving, it happens that some functionality needs to be changed, replaced
+or dropped completely. Such situations are inevitable. To reduce negative
+impact on your code, we introduce the deprecation process described below.
+
+We are still working on improvement of the current situation to provide stable
+interfaces for people, so extensibility of the current solutions could be much
+easier. This is just on of those improvements.
+
+## What is covered by deprecation process in leapp
+
+In short, leapp entities that are supposed to be used by other developers.
+That means e.g.:
+
+1. Models
+1. Classes and function in shared libraries in leapp repository
+1. Public APIs
+1. Actors providing functionality that could be used by any developer (produce
+   or consume messages)
+
+In other words, private classes, private functions, anything in private
+libraries, could be modified or removed without the deprecation process. As well,
+it's possible we will need to change something (e.g. a behaviour of a function)
+that will not be possible to cover reasonably by the deprecation process (e.g.
+change output of the function...). This should happens in as low number of cases
+as possible, but still it could happen. To limit such problems (and even those
+with the deprecated content), we recommend people to use APIs as much
+as possible.
+
+## What does it mean that something is deprecated?
+
+When you deprecate something, the only thing that changes is that the
+deprecated entity is marked in the code as deprecated which can have
+additional impact,likemessages produced on the user's terminal,
+in the report, ... But the rest of the functionality is the same as before,
+until the entity is removed completely.
+
+## What is the deprecation process for leapp
+
+In case some public functions, models,... (let's say, basically whatever that
+is supposed to be used by other people) are to be removed for any
+reason, these need to be marked as deprecated before the removal (if possible).
+In our projects, the deprecation will be applied only for leapp entities
+that have been introduced in an official release (IOW, a functionality that
+has been merged into the upstream, but has been removed before the release
+or was marked as experimental all the time, is going to be removed without
```

Comment:
I still can't understand, why do you referring this official release. As a developer, I confused here. I don't know what is experimental all the time and I don't know is it included in the official release? Probably, what makes it clear to me if I read something like this:
1. Check if the entity you would like to deprecate existed before the last release (git checkout release/bump-version-XXXX and check if this entity still presents)
2. If no, then no deprecation needed, otherwise it is needed
3. Experimented functionality means the functionality is marked as experimntal decorator (just example, since I didn't seen it so far)

Reply:
No, you do not have to check whether the functionality has been part of the last release or not (git blame can tell you that very quickly. btw) you can just deprecate it always as you want. but usually it doesn't make sense to create deprecation for something, that people have no possibility to use right? Same for the experimental think. When something is experimental (or there is clear disclaimer that people should not use it, it's experimental, ...) people cannot expect we will apply the deprecation process. we can, but why, as they couldn't use it?

Reply:
resolved over the the call